Fix parseUri to sanitize urls containing ASCII newline or tab

beshrkayali opened the pull request:
This fixes `parseUri` function in `uri` lib to sanitize provided url string from ascii newline and tabs. The URL standard
from WHATWG specifies that a parser should strip all tabs and newlines from input: https://url.spec.whatwg.org/#concept-basic-url-parser

Example (on `1.4.6`):

```nim
import uri
var res = initUri()
parseUri("java\nscript:alert('bad')", res)
echo res.path
```

Will output:
```sh
java
script:alert('bad')
```

This is a security issue that was fixed in Python 3.9.5 a couple of days ago ([issue43882](https://bugs.python.org/issue43882)) and it could allow some forms of attacks.

WARNING: This PR has been reverted. See https://github.com/nim-lang/Nim/pull/19128

beshrkayali committed:
Fix parseUri to sanitize urls containing ASCII newline or tab

timotheecour commented on lib/pure/uri.nim:
```diff
@@ -277,6 +278,12 @@ func resetUri(uri: var Uri) =
+func removeUnsafeBytesFromUri(uri: string): string =
+  var u = uri
+  for b in unsafeUrlBytesToRemove:
+    u = replace(u, b)
```
inefficient, this allocates more than it needs to.
use `multiReplace`, or even better do it manually (most efficient):
```nim
func removeUnsafeBytesFromUri(uri: string): string =
  # xxx we need to add to strbasics an in-place version:
  # `removeInplace(result: var string, chars: set[char])`
  const unsafeUrlBytesToRemove = {'\t', '\r', '\n'}
  for c in uri:
    if c notin unsafeUrlBytesToRemove: result.add c
```

beshrkayali replied:
Would something like this work for a `removeInplace`?

```nim
func removeInPlace(result: var string, chars: set[char]) =
  ## Removes all characters from `chars` from string `s`
  ## (in-place).
  runnableExamples:
    var userInput = "line\t1\nline\t2"
    removeInPlace(u, {'\n', '\t'})
    doAssert userInput == "line1line2"

  if result.len == 0: return
  var i = 0
  while i < result.len:
    if result[i] in chars: result.delete(i, i)
    i += 1
```

timotheecour commented on lib/pure/uri.nim:
```diff
@@ -277,6 +278,12 @@ func resetUri(uri: var Uri) =
       f = false
 
+func removeUnsafeBytesFromUri(uri: string): string =
+  var u = uri
```
general comment: use `result` variable, no need for `var u = ... return u`

timotheecour commented on lib/pure/uri.nim:
```diff
@@ -289,6 +296,16 @@ func parseUri*(uri: string, result: var Uri) =
     assert res.scheme == "https"
     assert res.hostname == "nim-lang.org"
     assert res.path == "/docs/manual.html"
+
```
```suggestion

    res = initUri()
```
(resets it)

timotheecour commented on lib/pure/uri.nim:
```diff
@@ -289,6 +296,16 @@ func parseUri*(uri: string, result: var Uri) =
+    assert res.path == "/docs/alert('msg')/"
+    assert res.query == "query=value"
+    assert res.anchor == "fragment"
+
```
I'm of the opinion that we should raise by default instead of silently removing \r\n\t, see also https://bugs.python.org/msg392995 (or whole thread in https://bugs.python.org/issue43882)

how about adding an optional param `strict = true` ?

beshrkayali replied:
@timotheecour Yeah, that makes sense. I think it's reasonable to do it as you're suggesting, defaulting to strict and allowing user to handle exception themselves.

The way I understand it, it's also what the WHATWG URL standard suggests:

> 2. If input contains any ASCII tab or newline, validation error.
> 3. Remove all ASCII tab or newline from input.

---
>  A validation error does not mean that the parser terminates. Termination of a parser is always stated explicitly, e.g., through a return statement.
> It is useful to signal validation errors as error-handling can be non-intuitive, legacy user agents might not implement correct error-handling, and the intent of what is written might be unclear to other developers.

timotheecour commented on lib/pure/uri.nim:
```diff
@@ -289,6 +296,16 @@ func parseUri*(uri: string, result: var Uri) =
+    parseUri("https://nim-lang\n.org\t/docs\nalert('msg\r\n')/?query\n=\tvalue#frag\nment", res)
+    assert res.scheme == "https"
+    assert res.hostname == "nim-lang.org"
+    assert res.path == "/docs/alert('msg')/"
```
test fails

beshrkayali committed:
Fix ups based on review

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

beshrkayali commented:
Thanks for the review @timotheecour! I pushed some fixes to address some aspects.

dom96 commented on lib/pure/uri.nim:
```diff
@@ -54,6 +54,7 @@ type
 
   UriParseError* = object of ValueError
 
+const unsafeUrlBytesToRemove = {'\t', '\r', '\n'}
```
Suggestion: can you add a link to the WHATWG spec that describes these characters as a comment here? :)

dom96 commented on lib/pure/uri.nim:
```diff
@@ -289,6 +294,33 @@ func parseUri*(uri: string, result: var Uri) =
     assert res.hostname == "nim-lang.org"
     assert res.path == "/docs/manual.html"
+
+    res = initUri()
```
These tests are too exhaustive for "runnable examples". They should go into the relevant test file (IIRC it's called turi.nim). Please move it there.

You can add a simpler `strict` example in here, but keep it as succinct as possible, readers of documentation don't want to sift through many lines of code to figure out semantics.

beshrkayali replied:
Makes sense! I simplified it a bit and moved tests to `turi.nim`.

dom96 commented on lib/pure/uri.nim:
```diff
@@ -289,6 +294,33 @@ func parseUri*(uri: string, result: var Uri) =
+    for c in uri:
+      if c in unsafeUrlBytesToRemove: uriParseError("Invalid uri '$#'" % uri)
+
+  let uri = removeUnsafeBytesFromUri(uri)
```
Nit: to avoid the cost of `removeUnsafeBytesFromUri`, maybe only do it for the non-strict case?

beshrkayali replied:
@dom96 That sounds reasonable. Not sure if the way I did it now is sane though :)

dom96 reviewed (approved):
Overall looks good, thank you for this! Main objections I have are to the overly long runnable examples :)

timotheecour commented on lib/pure/uri.nim:
```diff
@@ -277,7 +278,11 @@ func resetUri(uri: var Uri) =
+  for c in uri:
+    if c notin unsafeUrlBytesToRemove: result.add c
+
+func parseUri*(uri: string, result: var Uri, strict: bool = true) =
```
```suggestion
func parseUri*(uri: string, result: var Uri, strict = true) =
```

timotheecour commented on lib/pure/uri.nim:
```diff
@@ -289,6 +294,33 @@ func parseUri*(uri: string, result: var Uri) =
+
+    # Strict
+    doAssertRaises(UriParseError):
+      parseUri("https://nim-lang\n.org\t/docs/\nalert('msg\r\n')/?query\n=\tvalue#frag\nment", res, strict=true)
```
it's implicit in the declaration that strict defaults to true, so you can remove these few lines given that you're testing below with the `# Strict by default` block.

timotheecour commented on lib/pure/uri.nim:
```diff
@@ -325,7 +357,7 @@ func parseUri*(uri: string, result: var Uri) =
   parsePath(uri, i, result)
 
-func parseUri*(uri: string): Uri =
+func parseUri*(uri: string, strict: bool = true): Uri =
```
```suggestion
func parseUri*(uri: string, strict = true): Uri =
```

beshrkayali committed:
Additional fix ups based on review

- Avoid unnecessary `removeUnsafeBytesFromUri` call if parseUri is strict
- Move some parseUri tests to uri module test file

Co-authored-by: Dominik Picheta <dominikpicheta@googlemail.com>

beshrkayali commented:
@dom96 Thanks for the review! Pushed some changes to move tests, more concise runnableExamples, and avoid the unnecessary call to `removeUnsafeBytesFromUri` in strict case.

dom96 commented:
Awesome, thank you! Oh and one more thing, can you add this to the changelog? :)

beshrkayali committed:
Update changelog

beshrkayali commented:
@dom96 should be done now :)

dom96 commented:
Merging for now, but we might wish to be explicit about the semantics changing in the changelog (we didn't used to raise exceptions)

Araq commented:
> Merging for now, but we might wish to be explicit about the semantics changing in the changelog (we didn't used to raise exceptions)

That means it's a breaking change, we need to find a better solution.
